devicetypes/smartthings/zwave-lock.src/zwave-lock.groovy

```diff
@@ -1365,6 +1874,568 @@ def deleteCode(codeID) {
 	], 4200)
 }
 
+/**
+ * API end-point for updating the 'lockCodesData' attribute with the user info
+ *
+ * @param userInfo: Map consisting values for code id, phone number, access type, and last usage
+ */
+def setUserInfo(userInfo) {
+	log.trace "[DTH] Executing 'setUserInfo()' by ${device.displayName} with userInfo := ${userInfo}"
+	def codeID = userInfo.codeID
+	if (codeID) {
+		codeID = codeID.toString()
+		def lockCodesData = loadlockCodesData(codeID)
+		lockCodesData[codeID].lastUsage = userInfo.lastUsage
+		lockCodesData[codeID].userInfo.phoneNum = userInfo.phoneNum ?: ""
+		def accessType = userInfo.accessType
+		if (accessType) {
+			if (accessType == "always") {
+				lockCodesData[codeID].scheduleInfo = [:]
+				lockCodesData[codeID].scheduleInfo.scheduleData = [:]
+			}
+			lockCodesData[codeID].scheduleInfo.scheduleType = accessType
+		}
+		sendEvent(lockCodesDataEvent(lockCodesData))
+	}
+}
+
+/**
+ * Starts the scanning of schedule for the first code present in the 'lockCodes' map
+ */
+def startScanningSchedule() {
+	log.trace "[DTH] Executing 'startScanningSchedule()' by ${device.displayName}"
+	def codesToScanSchedules = state.codesToScanSchedules
+	def codeID = -1
+	if (codesToScanSchedules && codesToScanSchedules.size() > 0) {
+		// Schedule scanning did not complete the last time. Scan the remaining schedules.
+		codeID = codesToScanSchedules[0]
+	} else {
+		def lockCodes = loadLockCodes()
+		if (lockCodes && lockCodes.size() > 0) {
+			def keySet = lockCodes.keySet()
+			codesToScanSchedules = []
+			// Saving the list of lock codes for which the schedule scanning should be executed
+			// In case the user creates schedule while the scanning is in progress, those codes need not be scanned
+			for (def i = 1; i <= MAX_CODES_TO_SCAN; i++) {
+				if (i.toString() in keySet) {
+					codesToScanSchedules << i
+				}
+			}
+			state.codesToScanSchedules = codesToScanSchedules
```

Comment:
In what cases would this not be equivalent to `lockCodes.keySet()`?

Reply:
We scan only the first 8 slots when the lock is paired freshly. Among the first 8 slots it is possible that only few slots have lock code for e.g. only slot 1, 3, 4, 8. In this case the `codesToScanSchedules` list will contain 1, 3 ,4, 8 and for these slots schedules will be scanned. Meanwhile a user can create lock code in other available slots, either from smartapp or directly on the lock. In this case the new slots will get added to `lockCodes` map but we shouldn't be scanning those slots.

Reply:
how about 

`codesToScanSchedules = keySet.findall {it <= MAX_CODES_TO_SCAN}`